### ISSUE TITLE
Non-threadsafe mutable member variables

### ISSUE THREAD

**Issue by nilsdeppe** [sxs-collaboration/spectre#3936]
# Bug reports:

### Expected behavior:

Pretty much all of our code should be threadsafe. Only exception I think we have explicitly decided on is HDF5, and this is because we want to avoid being tied to a thtreadsafe HDF5. Threadsafe HDF5 is incompatible with parallel HDF5 (MPI-based).

### Current behavior:

In #3928 the `mutable` keyword started being linted for. We need to make sure that "all" places that use mutable members are switched to threadsafe. From that PR:

1. `DataBox` this is worrisome, but we'll need to look at where we have locks in the node groups. I think only in node groups can this be an issue. Compute tags are inherently not threadsafe, and it's not obvious there's an easy way to make them threadsafe. @kidder and I think it's fine to just say DataBox isn't threadsafe under some conditions, detailing the conditions in the documentation.
2. CCE: Not obviously bad since CCE generally operates on a single thread. However, we should check `Cce/WorldtubeDataManager.hpp` to make that's used in a single-threaded context only. The RobinsonTrautman solution is probably fine? Having said that, it might be fairly easy to get rid of the mutable member variables and then we don't have to worry about it.
3. `IO/H5/File.hpp`: since we don't require threadsafe HDF5, all HDF5 operations need to be explicitly threadsafe anyway. Should just add a `\warning` to the documentation.
4. Everything in `NumericalAlgorithms` should be threadsafe.
   YlmSpherepack is being fixed in #3931 and #934
   @nilsvu is probably best suited to fix the linear solvers
   Someone should fix `NumericalAlgorithms/Spectra/SwshInterpolation.hpp`
5. I don't know about `src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp`, should at least be documented to be non-threadsafe if that's the desired behavior.
6. `src/Time/BoundaryHistory.hpp` not threadsafe, but that should be fine. Should make sure it's documented as non-threadsafe.
7. `tests/Unit/Elliptic/SubdomainPreconditioners/Test_MinusLaplacian.cpp` and `tests/Unit/Helpers/NumericalAlgorithms/LinearSolver/TestHelpers.hpp` are probably fine since the tests are run in serial.

**Comment by nilsvu** [comment]
Yep I'll look at everything elliptic-related. At first glance those seem not obviously broken since they all run on a per-element basis (Schwarz solver, GMRES, ExplicitInverse, MinusLaplacian). Still, I'll look at them in detail and see if they can be made threadsafe or else document what's going on.
